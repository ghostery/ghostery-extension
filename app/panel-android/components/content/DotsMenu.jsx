/**
 * DotsMenu Component
 *
 * Ghostery Browser Extension
 * https://www.ghostery.com/
 *
 * Copyright 2018 Ghostery, Inc. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0
 */

import React from 'react';
import PropTypes from 'prop-types';

export default class DotsMenu extends React.Component {
	constructor(props) {
		super(props);

		this.state = {
			opening: false,
		};
	}

	componentDidMount() {
		window.addEventListener('click', this.handleClick, false);
	}

	componentWillUnmount() {
		window.removeEventListener('click', this.handleClick, false);
	}

	/* Close the menu if user clicks anywhere on the window */
	handleClick = (event) => {
		if (this.state.opening && event.target.className.indexOf('dots-menu-btn') === -1) {
			this.setState({
				opening: false,
			});
		}
	}

	/* Toggle menu */
	dotsButtonClicked = () => {
		const currentState = this.state.opening;

		this.setState({
			opening: !currentState,
		});
	}

	render() {
		return (
			<div className="dots-menu">
				<button className="dots-menu-btn" onClick={this.dotsButtonClicked} />
				<div className={`dots-menu-content ${this.state.opening ? 'opening' : ''}`}>
					<ul>
						{this.props.actions.map(action =>
							(
								<li key={action.id}>
									<button className="dots-menu-item" onClick={action.callback}>{action.name}</button>
								</li>
							)
						)}
					</ul>
				</div>
			</div>
		);
	}
}

DotsMenu.propTypes = {
	actions: PropTypes.arrayOf(PropTypes.object).isRequired,
};
