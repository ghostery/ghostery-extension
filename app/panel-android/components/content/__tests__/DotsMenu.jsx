/**
 * DotsMenu Test Component
 *
 * Ghostery Browser Extension
 * https://www.ghostery.com/
 *
 * Copyright 2020 Ghostery, Inc. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0
 */

import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import DotsMenu from '../DotsMenu';

describe('app/panel-android/components/content/DotsMenu.jsx', () => {
	describe('Snapshot tests with react-test-renderer', () => {
		test('DotsMenu component with 0 actions', () => {
			const actions = [];
			const component = renderer.create(
				<DotsMenu actions={actions} />
			).toJSON();
			expect(component).toMatchSnapshot();
		});

		test('DotsMenu component with 3 actions', () => {
			const actions = [
				{
					id: 'action-1',
					name: 'Action One',
					callback: () => {},
				},
				{
					id: 'action-2',
					name: 'Action Two',
					callback: () => {},
				},
				{
					id: 'action-3',
					name: 'Action Three',
					callback: () => {},
				}
			];

			const component = renderer.create(
				<DotsMenu actions={actions} />
			).toJSON();
			expect(component).toMatchSnapshot();
		});
	});

	describe('Functionality tests mounted with Enzyme', () => {
		test('DotsMenu component with 3 actions happy path', () => {
			const actions = [
				{
					id: 'action-1',
					name: 'Action One',
					callback: jest.fn(),
				},
				{
					id: 'action-2',
					name: 'Action Two',
					callback: jest.fn(),
				},
				{
					id: 'action-3',
					name: 'Action Three',
					callback: jest.fn(),
				}
			];

			const component = mount(
				<DotsMenu actions={actions} />
			);
			expect(component.find('.DotsMenu').length).toBe(1);
			expect(component.find('.DotsMenu__button').length).toBe(1);
			expect(component.find('.DotsMenu__content').length).toBe(1);
			expect(component.find('.DotsMenu__content.DotsMenu__open').length).toBe(0);
			expect(component.find('.DotsMenu__item').length).toBe(3);

			component.setState({ open: true });
			expect(component.find('.DotsMenu__content.DotsMenu__open').length).toBe(1);
			expect(component.find('.DotsMenu__item').length).toBe(3);
			expect(actions[0].callback.mock.calls.length).toBe(0);
			expect(actions[1].callback.mock.calls.length).toBe(0);
			expect(actions[2].callback.mock.calls.length).toBe(0);
			component.find('.DotsMenu__item').at(0).simulate('click');
			expect(actions[0].callback.mock.calls.length).toBe(1);
			component.find('.DotsMenu__item').at(1).simulate('click');
			expect(actions[1].callback.mock.calls.length).toBe(1);
			component.find('.DotsMenu__item').at(2).simulate('click');
			expect(actions[0].callback.mock.calls.length).toBe(1);
			expect(actions[1].callback.mock.calls.length).toBe(1);
			expect(actions[2].callback.mock.calls.length).toBe(1);

			component.setState({ open: false });
			expect(component.find('.DotsMenu__content.DotsMenu__open').length).toBe(0);
			expect(component.find('.DotsMenu__item').length).toBe(3);
		});
	});
});
