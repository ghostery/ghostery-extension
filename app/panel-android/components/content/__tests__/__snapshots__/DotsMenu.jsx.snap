// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/panel-android/components/content/DotsMenu.jsx Snapshot tests with react-test-renderer DotsMenu component with 0 actions 1`] = `
<div
  className="DotsMenu"
>
  <button
    aria-label="Menu"
    className="DotsMenu__button"
    onClick={[Function]}
    type="button"
  />
  <div
    className="DotsMenu__content"
  >
    <ul />
  </div>
</div>
`;

exports[`app/panel-android/components/content/DotsMenu.jsx Snapshot tests with react-test-renderer DotsMenu component with 3 actions 1`] = `
<div
  className="DotsMenu"
>
  <button
    aria-label="Menu"
    className="DotsMenu__button"
    onClick={[Function]}
    type="button"
  />
  <div
    className="DotsMenu__content"
  >
    <ul>
      <li>
        <button
          className="DotsMenu__item"
          onClick={[Function]}
          type="button"
        >
          Action One
        </button>
      </li>
      <li>
        <button
          className="DotsMenu__item"
          onClick={[Function]}
          type="button"
        >
          Action Two
        </button>
      </li>
      <li>
        <button
          className="DotsMenu__item"
          onClick={[Function]}
          type="button"
        >
          Action Three
        </button>
      </li>
    </ul>
  </div>
</div>
`;
